src/controllers/legacycontrollersettingsfactory.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <QDomElement>
+#include <QLabel>
+#include <QLayout>
+#include <QList>
+#include <QObject>
+#include <QSizePolicy>
+#include <QSpinBox>
+#include <QString>
+#include <QWidget>
+#include <limits>
+#include <memory>
+#include <tuple>
+
+#include "controllers/legacycontrollersettingsfactory.h"
+
+class AbstractLegacyControllerSetting;
+
+/// @brief This class defines an interface that a controller setting type must
+/// implement so it can be used properly by the builder
+/// @tparam T The class implementing this interface
+template<class T>
+class LegacyControllerSettingFactory {
+    inline static LegacyControllerSettingFactory* createFrom(const QDomElement& element) {
+        return new T(element);
+    }
+    inline static bool match(const QDomElement& element) {
+        return T::match(element);
+    }
+};
+
+/// @brief This class is used to dynamically instantiate a controller setting based on its type
+class LegacyControllerSettingBuilder {
+  public:
+    static LegacyControllerSettingBuilder* instance();
+
+    /// @brief Register a new type of setting. This method is used by the
+    /// REGISTER macro, it shouldn't be used directly
+    /// @param match the match function of the new setting
+    /// @param creator the creator function of the new setting
+    /// @return Always true
+    bool registerType(bool (*match)(const QDomElement&),
```

Comment:
This can become a template, to get rid of the REGISTER_LEGACY_CONTROLLER_SETTING macro

Reply:
I'm not sure how this would replace the type registration, can you show an example of what you mean? Just top confirm, `REGISTER_LEGACY_CONTROLLER_SETTING` follows the same patter than [Q_DECLARE_METATYPE](https://doc.qt.io/qt-6/qmetatype.html#Q_DECLARE_METATYPE) - it will generate some "dummy" code which will allow dynamic registration of meta information about a type (`LegacyControllerSettingBuilder` instead of `QObject` in the present case)

Reply:
`LegacyControllerSettingBuilder::instance()->registerType(__VA_ARGS__::match, __VA_ARGS__::createFrom)`
for
`REGISTER_LEGACY_CONTROLLER_SETTING(LegacyControllerBooleanSetting);`
can become 
`LegacyControllerSettingBuilder::instance()->registerType<LegacyControllerBooleanSetting>()`

Than we have here (untested)
```
template<class T>
bool registerType() {
    m_supportedSettings.append(std::make_tuple(
            &T::match, 
            &T::creator));
    return true;
}
```